src/connector/src/base.rs

```diff
@@ -47,22 +52,93 @@ pub struct SourceMessage {
 }
 
 /// The metadata of a split.
-pub trait SourceSplit: Sized {
+pub trait SplitMetaData: Sized {
     fn id(&self) -> String;
-    fn to_string(&self) -> Result<String>;
+    fn to_json_bytes(&self) -> Bytes;
     fn restore_from_bytes(bytes: &[u8]) -> Result<Self>;
 }
 
 /// The persistent state of the connector.
-#[derive(Debug, Clone)]
+#[derive(Debug, Clone, Serialize, Deserialize)]
 pub struct ConnectorState {
     pub identifier: Bytes,
     pub start_offset: String,
     pub end_offset: String,
 }
 
+impl ConnectorState {
+    pub fn from_hashmap(state: HashMap<String, String>) -> Vec<Self> {
+        if state.is_empty() {
+            return vec![];
+        }
+        let mut connector_states: Vec<Self> = Vec::with_capacity(state.len());
+        for (split, offset) in state {
+            connector_states.push(Self {
+                identifier: Bytes::from(split),
+                start_offset: offset.clone(),
+                end_offset: "".to_string(),
+            })
+        }
+        connector_states
```

Comment:
```suggestion
        let mut connector_states: Vec<Self> = Vec::with_capacity(state.len());
        connector_states.extend(state.iter().map(|(split, offset)| {
            Self {
                identifier: Bytes::from(split),
                start_offset: offset.clone(),
                end_offset: "".to_string(),
            }
        }))
        connector_states
```

Use extend whenever it's possible.